interfaces/apparmor/backend_test.go

```diff
@@ -492,7 +495,362 @@ func (s *backendSuite) TestSetupHostSnapConfineApparmorForReexecWritesNew(c *C)
+		// Errors from parsing mountinfo and fstab are propagated.
+		{
+			mountinfo:    "bad syntax",
+			errorPattern: "cannot parse /proc/self/mountinfo, .*",
+		},
+		{
+			fstab:        "bad syntax",
+			errorPattern: "cannot parse /etc/fstab, .*",
+		},
+		// NFSv4 currently mounted at /home/zyga/nfs is recognized.
+		{
+			mountinfo: "680 27 0:59 / /home/zyga/nfs rw,relatime shared:478 - nfs4 localhost:/srv/nfs rw,vers=4.2,rsize=524288,wsize=524288,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1",
+			nfs:       true,
+		},
+		// NFSv4 currently mounted at /mnt/nfs is ignored (not in $HOME).
+		{
+			mountinfo: "680 27 0:59 / /mnt/nfs rw,relatime shared:478 - nfs4 localhost:/srv/nfs rw,vers=4.2,rsize=524288,wsize=524288,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1",
+		},
+		// NFSv4 that may be mounted at /home/zyga/nfs is recognized.
+		{
+			fstab: "localhost:/srv/nfs /home/zyga/nfs nfs defaults 0 0",
+			nfs:   true,
+		},
+		// NFSv4 that may be mounted at /mnt/nfs is ignored (not in $HOME).
+		{
+			fstab: "localhost:/srv/nfs /mnt/nfs nfs defaults 0 0",
+		},
+	}
+	for _, tc := range cases {
+		restore := apparmor.MockMountInfo(tc.mountinfo)
+		defer restore()
+		restore = apparmor.MockEtcFstab(tc.fstab)
+		defer restore()
+
+		nfs, err := apparmor.IsHomeUsingNFS()
+		if tc.errorPattern != "" {
+			c.Assert(err, ErrorMatches, tc.errorPattern, Commentf("test case %#v", tc))
+		} else {
+			c.Assert(err, IsNil)
+		}
+		c.Assert(nfs, Equals, tc.nfs)
+	}
+}
+
+// snap-confine policy when NFS is not used.
+func (s *backendSuite) TestSetupSnapConfineGeneratedPolicyNoNFS(c *C) {
+	// Make it appear as if NFS was not used.
+	restore := apparmor.MockMountInfo("")
+	defer restore()
+	restore = apparmor.MockEtcFstab("")
+	defer restore()
+
+	// Intercept interaction with apparmor_parser
+	cmd := testutil.MockCommand(c, "apparmor_parser", "")
+	defer cmd.Restore()
+
+	// Setup generated policy for snap-confine.
+	err := apparmor.SetupSnapConfineGeneratedPolicy()
+	c.Assert(err, IsNil)
+	c.Assert(cmd.Calls(), HasLen, 0)
+
+	// Because NFS is not used there are no local policy files but the
+	// directory was created.
+	files, err := ioutil.ReadDir(dirs.SnapConfineAppArmorDir)
+	c.Assert(err, IsNil)
+	c.Assert(files, HasLen, 0)
+
+	// The policy was not reloaded.
+	c.Assert(cmd.Calls(), HasLen, 0)
+}
+
+// snap-confine policy when NFS is used and snapd has not re-executed.
+func (s *backendSuite) TestSetupSnapConfineGeneratedPolicyWithNFS(c *C) {
+	// Make it appear as if NFS was mounted under /home.
+	restore := apparmor.MockMountInfo("680 27 0:59 / /home/zyga/nfs rw,relatime shared:478 - nfs4 localhost:/srv/nfs rw,vers=4.2,rsize=524288,wsize=524288,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1")
+	defer restore()
+	restore = apparmor.MockEtcFstab("")
```

Comment:
I noticed that we have a lot of tests using `MockMountInfo("...")` when using `MockEtcFstab("")`, but not the other way around or when both are set (outside of `TestIsHomeUsingNFS()`). Was this intentional?

Reply:
I wrote mountinfo support code first and only then added the trigger code that looked for fstab. Since other tests are checking that they have equivalent power I should perhaps just say "mock stuff so that NFS workaround is enabled" without going into the details of how that is done.

Reply:
Also done now.